workflow/controller/exec_control.go

```diff
@@ -75,7 +75,7 @@ func (woc *wfOperationCtx) applyExecutionControl(pod *apiv1.Pod, wfNodesLock *sy
 	var newDeadline *time.Time
 	if woc.wf.Spec.Shutdown != "" {
 		_, onExitPod := pod.Labels[common.LabelKeyOnExit]
-		if woc.wf.Spec.Shutdown == wfv1.ShutdownStrategyTerminate || (woc.wf.Spec.Shutdown == wfv1.ShutdownStrategyStop && !onExitPod) {
+		if !woc.wf.Spec.Shutdown.ShouldExecute(onExitPod) {
```

Comment:
This is unrelated to the main PR. It simply uses a utility function instead of dealing with the logic manually